Detect bad unstructured conversions

wojtek-t opened the pull request:
Ref https://github.com/kubernetes/kubernetes/issues/39017

This PR also speed up the conversion:
before:
```
BenchmarkToFromUnstructured-12           	    1000	   1201132 ns/op	   15335 B/op	     268 allocs/op
BenchmarkToFromUnstructuredViaJSON-12    	    1000	   2127384 ns/op	   29669 B/op	     457 allocs/op
```
after:
```
BenchmarkToFromUnstructured-12           	    2000	    911243 ns/op	   10472 B/op	     196 allocs/op
BenchmarkToFromUnstructuredViaJSON-12    	    1000	   2243216 ns/op	   29665 B/op	     457 allocs/op
```

wojtek-t committed:
Minor cleanup in unstructured converter

wojtek-t committed:
Detection of failed unstructured conversions.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41401)
<!-- Reviewable:end -->

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:
```diff
@@ -43,13 +43,13 @@ type B struct {
-	G []int `json"cg"`
-	H float32 `json:ch"`
+	G []int64 `json:"cg"`
+//	H float32 `json:"ch"`
```
Because of some reason, floats translate differently in custom code and when using json marshaling - need to figure out why.

smarterclayton replied:
@liggitt knows all the pain of floats and JSON

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -465,6 +524,25 @@ func pointerToUnstructured(sv, dv reflect.Value) error {
+	case reflect.Float32, reflect.Float64:
+		return v.Float() == 0
+	case reflect.Map, reflect.Slice, reflect.Ptr, reflect.Interface:
+		// TODO: It seems that 0-len maps are ignored in it.
```
Some tests are failing when checking "object -> map[string]interface{}" conversion is on.

It seems related to empty maps, but didn't yet figure it out.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -481,17 +559,21 @@ func structToUnstructured(sv, dv reflect.Value) error {
+			// No fource field, skip.
+			continue
+		}
+		if fieldInfo.omitempty && isZero(fv) {
```
@smarterclayton - this is the key optimization that we were missing previously (and that is widely used in our objects).

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -481,17 +559,21 @@ func structToUnstructured(sv, dv reflect.Value) error {
 		fv := sv.Field(i)
 
+		if !fv.IsValid() {
+			// No fource field, skip.
```
source?

smarterclayton replied:
How can fv be invalid here?

wojtek-t replied:
It was from my experiments - I don't think it's needed.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -304,8 +344,27 @@ func interfaceFromUnstructured(sv, dv reflect.Value) error {
+		newUnstr := &map[string]interface{}{}
+		newErr := toUnstructuredViaJSON(obj, newUnstr)
+		if (err != nil) != (newErr != nil) {
+			glog.Fatalf("ToUnstructured unexpected error: %v, expected: %v", err, newErr)
```
better to panic, include the type information of the new and old.  We'll want to know what code path led to this.

wojtek-t replied:
I don't understand this comment- glog.Fatalf also gives the stacktrace.

I also don't understand the second part of the comment - if one is error and the other is not, objects doesn't make much sense (at least in case of those that returned error).

smarterclayton replied:
Fatal doesn't give the call stack - it calls os.Exit().

Error messages (as we have them) are fairly bad at providing the exact context for the error.  If we have this on and are debugging it in a large scale test, I want all the possible info contained in the output.

wojtek-t replied:
Hmm - maybe glog.Fatalf changed recently? I'm clearly getting a stacktrace:
```
wojtekt@wojtekt-work:~/go/src/k8s.io/kubernetes$ hack/test-go.sh k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured/
NOTE: hack/test-go.sh has been replaced by 'make test'

The equivalent of this invocation is: 
    make test WHAT='k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured/'


Running tests for APIVersion: v1,apps/v1beta1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2alpha1,batch/v1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,policy/v1beta1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,storage.k8s.io/v1beta1,federation/v1beta1
+++ [0215 09:20:21] Running tests without code coverage
F0215 09:20:23.022998   16627 converter.go:375] ToUnstructured mismatch: &map[string]interface {}{"ba":map[string]interface {}{}, "bd":interface {}(nil), "ce":interface {}(nil), "cf":interface {}(nil), "ci":interface {}(nil), "ca":interface {}(nil), "bc":interface {}(nil), "cd":interface {}(nil), "cg":interface {}(nil), "ch":0}, diff: 
object[bb]:
  a: <nil>
  b: ""
object[cc]:
  a: <nil>
  b: "ccc"
object[ch]:
  a: 0
  b: 0
goroutine 19 [running]:
k8s.io/kubernetes/vendor/github.com/golang/glog.stacks(0xb48200, 0xc400000000, 0x1ad, 0x1fc)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:766 +0xa5
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).output(0xb2ad20, 0xc400000003, 0xc420032540, 0xaddddf, 0xc, 0x177, 0x0)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:717 +0x337
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).printf(0xb2ad20, 0xc400000003, 0x8cfbdc, 0x26, 0xc420145c60, 0x2, 0x2)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:655 +0x14c
k8s.io/kubernetes/vendor/github.com/golang/glog.Fatalf(0x8cfbdc, 0x26, 0xc420145c60, 0x2, 0x2)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:1145 +0x67
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured.(*ConverterImpl).ToUnstructured(0xc4203db629, 0xae34a0, 0xc42001a7e0, 0xc4204d82c0, 0xc4202f8f90, 0x0)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:375 +0x39d
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured.doRoundTrip(0xc4200323c0, 0xae34a0, 0xc42001a7e0)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:125 +0x5f3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured.TestRoundTrip(0xc4200323c0)
	/usr/local/google/home/wojtekt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:216 +0x946
testing.tRunner(0xc4200323c0, 0x901360)
	/usr/local/google/home/wojtekt/.gvm/gos/go1.7.4/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/google/home/wojtekt/.gvm/gos/go1.7.4/src/testing/testing.go:646 +0x2ec
FAIL	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured	0.211s
make: *** [test] Error 1
```

Regarding errors - ok, got your point. Fixed.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -481,17 +559,21 @@ func structToUnstructured(sv, dv reflect.Value) error {
 		fieldInfo := fieldInfoFromField(st, i)
 		fv := sv.Field(i)
 
+		if !fv.IsValid() {
```
I missed this is the earlier review, but if the field name is `-` the field should be skipped.  I don't see that in the code, but we use it in a few places.

wojtek-t replied:
Added

wojtek-t commented:
@smarterclayton - It's still not fully ready for review, but I fixed a bunch of issues. And at the same time, this is faster now (and less allocations) - I updated the PR description.

wojtek-t commented:
@smarterclayton - this should now work. PTAL

wojtek-t reviewed (commented):
@smarterclayton - PTAL

wojtek-t commented:
@smarterclayton friendly ping

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -57,18 +57,20 @@ var (
+// ConverterImpl knows how to convert betweek runtime.Object and
 // Unstructured in both ways.
-type Converter struct {
+type ConverterImpl struct {
```
Make this private, and the constructer public, and expose a public interface.  Basically hide impl and expose interface and constructor.

wojtek-t replied:
Fixed

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go:
```diff
@@ -64,14 +76,40 @@ var (
+	// If true, we will be additionally running conversion via json
+	// to ensure that the result is true.
+	// This is supposed to be set only in tests.
+	mismatchDetection bool
```
Would prefer this be an env var setup to instantiate the default instance:

```
var DefaultConvertor = NewConvertor(parsebool(os.Getenv(...))
```

wojtek-t replied:
Fixed

smarterclayton commented:
Two comments to have a consistent setup / constructor.  Everything else looks fine.

wojtek-t committed:
Unified unstructured converter interface and setup

wojtek-t commented:
@smarterclayton - comments applied. PTAL

wojtek-t commented:
Fixed gofmt.

smarterclayton commented:
/lgtm

smarterclayton commented:
This flake link
<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41401/pull-kubernetes-node-e2e-non-cri/164/>
looks
non-related but I see no other references to it.

@k8s-bot non-cri node e2e test this

On Fri, Feb 17, 2017 at 4:07 PM, k8s-ci-robot <notifications@github.com>
wrote:

> @wojtek-t <https://github.com/wojtek-t>: The following test(s) *failed*:
> Test name Commit Details Rerun command
> Jenkins non-CRI GCE Node e2e 924e9a5
> <https://github.com/kubernetes/kubernetes/commit/924e9a5b3aad8df9b162e3acd2ef0dd2fbf2ffce>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41401/pull-kubernetes-node-e2e-non-cri/164/> @k8s-bot
> non-cri node e2e test this
>
> Full PR test history <https://k8s-gubernator.appspot.com/pr/41401>. Your
> PR dashboard <https://k8s-gubernator.appspot.com/pr/wojtek-t>. Please
> help us cut down on flakes by linking to an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://github.com/kubernetes/community/blob/master/contributors/devel/pull-request-commands.md>.
> If you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://github.com/kubernetes/test-infra/blob/master/prow/commands.md>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41401#issuecomment-280766588>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p1f_NOKOBX4KHNEAK6Zv20eKycljks5rdgwDgaJpZM4MATJi>
> .
>
